File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/anthropic/claude-opus-4.8_tooling=none/rep1/brazilian_soccer_mcp/queries.py
"""
Context
=======
Module: brazilian_soccer_mcp.queries

Pure query/aggregation layer over a `KnowledgeBase`. The MCP server (server.py)
is a thin adapter that calls these functions and formats their dict results;
keeping the logic here means it is directly unit-testable without the protocol
in the way (see the BDD tests).

Every public function:
  * accepts a `KnowledgeBase` plus plain query parameters,
  * matches team/player names via the canonical keys precomputed at load time
    (`normalize.names_match` semantics, vectorised here for speed),
  * returns JSON-serialisable Python (dicts/lists/primitives) - never raw
    DataFrames - so the server can hand results straight to an LLM.

Design choices worth knowing:
  * Standings/records exclude matches with missing goals (the 24 NA rows),
    because you cannot award points for a result you don't have.
  * Head-to-head and team records are computed from the home/away perspective
    in one pass over the (already small) filtered frame.
  * "Brasileirão" as a competition filter means Série A specifically - that is
    what users mean by the league title; Série B/C are addressable by their
    full names.
"""

from __future__ import annotations

import pandas as pd

from .data_loader import SERIE_A, KnowledgeBase
from .normalize import canonical

# User-facing competition aliases -> stored competition label.
_COMP_ALIASES = {
    "brasileirao": SERIE_A,
    "brasileirao serie a": SERIE_A,
    "serie a": SERIE_A,
    "serie b": "Brasileirão Série B",
    "serie c": "Brasileirão Série C",
    "copa do brasil": "Copa do Brasil",
    "brazilian cup": "Copa do Brasil",
    "libertadores": "Copa Libertadores",
    "copa libertadores": "Copa Libertadores",
}


def resolve_competition(name: str | None) -> str | None:
    """Map a free-text competition name to a stored label, or None if blank.

    Raises ValueError on an unrecognised non-empty value so callers can report
    the supported set rather than silently returning everything.
    """
    if name is None or str(name).strip() == "":
        return None
    key = canonical(name)
    if key in _COMP_ALIASES:
        return _COMP_ALIASES[key]
    # Allow passing the exact stored label through unchanged.
    raise ValueError(
        f"Unknown competition '{name}'. Known: Brasileirão Série A/B/C, "
        "Copa do Brasil, Copa Libertadores."
    )


def _team_mask(df: pd.DataFrame, column: str, team: str) -> pd.Series:
    """Vectorised canonical substring match against a *_canon column."""
    key = canonical(team)
    col = df[column].fillna("")
    if not key:
        return pd.Series(False, index=df.index)
    # Either side may contain the other (see normalize.names_match): exact
    # match, stored key contains the query, or query contains the stored key.
    contains_key = col.str.contains(key, regex=False)
    key_contains = col.apply(lambda c: bool(c) and c in key)
    return col.eq(key) | contains_key | key_contains


def _apply_filters(
    kb: KnowledgeBase,
    competition: str | None = None,
    season: int | None = None,
) -> pd.DataFrame:
    df = kb.matches
    if competition is not None:
        df = df[df["competition"] == resolve_competition(competition)]
    if season is not None:
        df = df[df["season"] == int(season)]
    return df


def _match_record(row: pd.Series) -> dict:
    """Serialise one match row to a compact, display-ready dict."""
    d = row["match_date"]
    hg = row["home_goal"]
    ag = row["away_goal"]
    return {
        "date": d.isoformat() if d is not None and not pd.isna(d) else None,
        "competition": row["competition"],
        "season": int(row["season"]) if pd.notna(row["season"]) else None,
        "round": None if pd.isna(row["round"]) else str(row["round"]),
        "stage": None if pd.isna(row["stage"]) else str(row["stage"]),
        "home": row["home"],
        "away": row["away"],
        "home_goal": int(hg) if pd.notna(hg) else None,
        "away_goal": int(ag) if pd.notna(ag) else None,
        "score": (
            f"{int(hg)}-{int(ag)}" if pd.notna(hg) and pd.notna(ag) else None
        ),
        "source": row["source"],
    }


def _sort_by_date(df: pd.DataFrame, ascending: bool = True) -> pd.DataFrame:
    # NaT dates sort last regardless of direction so dated rows lead.
    return df.sort_values("match_date", ascending=ascending, na_position="last")


# --------------------------------------------------------------------------- #
# 1. Match queries
# --------------------------------------------------------------------------- #
def find_matches(
    kb: KnowledgeBase,
    team: str | None = None,
    opponent: str | None = None,
    competition: str | None = None,
    season: int | None = None,
    venue: str = "either",
    limit: int = 50,
) -> dict:
    """Find matches by team(s), competition, season, and home/away venue.

    venue: 'home' | 'away' | 'either' (interpreted relative to `team`).
    When `opponent` is given, returns only matches between the two teams
    (in either ground). Results are date-sorted (most recent first).
    """
    df = _apply_filters(kb, competition, season)

    if team:
        home_hit = _team_mask(df, "home_canon", team)
        away_hit = _team_mask(df, "away_canon", team)
        if venue == "home":
            df = df[home_hit]
        elif venue == "away":
            df = df[away_hit]
        else:
            df = df[home_hit | away_hit]

    if opponent:
        opp_home = _team_mask(df, "home_canon", opponent)
        opp_away = _team_mask(df, "away_canon", opponent)
        df = df[opp_home | opp_away]

    total = len(df)
    rows = _sort_by_date(df, ascending=False).head(limit)
    return {
        "count": total,
        "returned": len(rows),
        "matches": [_match_record(r) for _, r in rows.iterrows()],
    }


# --------------------------------------------------------------------------- #
# 2. Team queries
# --------------------------------------------------------------------------- #
def _accumulate_record(df: pd.DataFrame, team: str) -> dict:
    """Compute W/D/L, goals for/against from a team's perspective.

    Matches missing goals are ignored. Operates over `df` already filtered to
    the team's matches.
    """
    home_mask = _team_mask(df, "home_canon", team)
    played = wins = draws = losses = gf = ga = 0
    for _, row in df.iterrows():
        hg, ag = row["home_goal"], row["away_goal"]
        if pd.isna(hg) or pd.isna(ag):
            continue
        is_home = bool(home_mask.loc[row.name])
        scored, conceded = (int(hg), int(ag)) if is_home else (int(ag), int(hg))
        played += 1
        gf += scored
        ga += conceded
        if scored > conceded:
            wins += 1
        elif scored < conceded:
            losses += 1
        else:
            draws += 1
    win_rate = round(100 * wins / played, 1) if played else 0.0
    return {
        "matches": played,
        "wins": wins,
        "draws": draws,
        "losses": losses,
        "goals_for": gf,
        "goals_against": ga,
        "goal_difference": gf - ga,
        "points": wins * 3 + draws,
        "win_rate": win_rate,
    }


def team_record(
    kb: KnowledgeBase,
    team: str,
    competition: str | None = None,
    season: int | None = None,
    venue: str = "either",
) -> dict:
    """Aggregate a team's record, optionally scoped by competition/season/venue."""
    df = _apply_filters(kb, competition, season)
    home_hit = _team_mask(df, "home_canon", team)
    away_hit = _team_mask(df, "away_canon", team)
    if venue == "home":
        scoped = df[home_hit]
    elif venue == "away":
        scoped = df[away_hit]
    else:
        scoped = df[home_hit | away_hit]

    if scoped.empty:
        return {
            "team": team,
            "found": False,
            "competition": competition,
            "season": season,
            "venue": venue,
            "record": _accumulate_record(scoped, team),
        }

    # Display name = most common stored spelling among the team's rows. Recompute
    # masks against `scoped` so indices always align with the filtered frame.
    sh = _team_mask(scoped, "home_canon", team)
    sa = _team_mask(scoped, "away_canon", team)
    names = pd.concat([scoped.loc[sh, "home"], scoped.loc[sa, "away"]])
    display = names.mode().iat[0] if not names.empty else team
    return {
        "team": display,
        "found": True,
        "competition": competition,
        "season": season,
        "venue": venue,
        "record": _accumulate_record(scoped, team),
    }


def head_to_head(
    kb: KnowledgeBase,
    team_a: str,
    team_b: str,
    competition: str | None = None,
    season: int | None = None,
) -> dict:
    """Head-to-head summary between two teams across the (filtered) data."""
    df = _apply_filters(kb, competition, season)
    a_home = _team_mask(df, "home_canon", team_a)
    a_away = _team_mask(df, "away_canon", team_a)
    b_home = _team_mask(df, "home_canon", team_b)
    b_away = _team_mask(df, "away_canon", team_b)
    h2h = df[(a_home & b_away) | (a_away & b_home)]

    a_wins = b_wins = draws = a_goals = b_goals = 0
    for _, row in h2h.iterrows():
        hg, ag = row["home_goal"], row["away_goal"]
        if pd.isna(hg) or pd.isna(ag):
            continue
        a_is_home = bool(a_home.loc[row.name])
        a_sc, b_sc = (int(hg), int(ag)) if a_is_home else (int(ag), int(hg))
        a_goals += a_sc
        b_goals += b_sc
        if a_sc > b_sc:
            a_wins += 1
        elif b_sc > a_sc:
            b_wins += 1
        else:
            draws += 1

    rows = _sort_by_date(h2h, ascending=False)
    return {
        "team_a": team_a,
        "team_b": team_b,
        "total_matches": len(h2h),
        "team_a_wins": a_wins,
        "team_b_wins": b_wins,
        "draws": draws,
        "team_a_goals": a_goals,
        "team_b_goals": b_goals,
        "matches": [_match_record(r) for _, r in rows.iterrows()],
    }


# --------------------------------------------------------------------------- #
# 3. Player queries
# --------------------------------------------------------------------------- #
def _player_record(row: pd.Series) -> dict:
    def g(col):
        if col not in row.index:
            return None
        v = row[col]
        return None if pd.isna(v) else v

    overall = g("Overall")
    age = g("Age")
    return {
        "id": int(g("ID")) if g("ID") is not None else None,
        "name": g("Name"),
        "age": int(age) if age is not None else None,
        "nationality": g("Nationality"),
        "overall": int(overall) if overall is not None else None,
        "potential": int(g("Potential")) if g("Potential") is not None else None,
        "club": g("Club"),
        "position": g("Position"),
    }


def find_players(
    kb: KnowledgeBase,
    name: str | None = None,
    nationality: str | None = None,
    club: str | None = None,
    position: str | None = None,
    min_overall: int | None = None,
    limit: int = 25,
) -> dict:
    """Search FIFA players by name/nationality/club/position/rating.

    All provided filters are ANDed. Results sort by Overall descending so the
    "top players" intent is the default ordering.
    """
    df = kb.players
    mask = pd.Series(True, index=df.index)
    if name:
        key = canonical(name)
        mask &= df["name_canon"].str.contains(key, regex=False)
    if nationality:
        key = canonical(nationality)
        mask &= df["nationality_canon"].eq(key) | df["nationality_canon"].str.contains(
            key, regex=False
        )
    if club:
        key = canonical(club)
        mask &= df["club_canon"].str.contains(key, regex=False)
    if position:
        pos = str(position).strip().upper()
        mask &= df["Position"].str.upper().eq(pos)
    if min_overall is not None:
        mask &= df["Overall"] >= int(min_overall)

    hits = df[mask].sort_values("Overall", ascending=False, na_position="last")
    total = len(hits)
    rows = hits.head(limit)
    return {
        "count": total,
        "returned": len(rows),
        "players": [_player_record(r) for _, r in rows.iterrows()],
    }


def players_by_club_summary(
    kb: KnowledgeBase, nationality: str | None = "Brazil", top_n: int = 10
) -> dict:
    """Group players (optionally by nationality) by club: count + avg rating.

    Answers "Brazilian players at Brazilian clubs" style asks. Clubs ordered by
    player count descending, then average rating.
    """
    df = kb.players
    if nationality:
        key = canonical(nationality)
        df = df[df["nationality_canon"].eq(key)]
    df = df[df["Club"].astype(str).str.len() > 0]
    grouped = (
        df.groupby("Club")
        .agg(players=("Name", "count"), avg_overall=("Overall", "mean"))
        .sort_values(["players", "avg_overall"], ascending=False)
        .head(top_n)
    )
    return {
        "nationality": nationality,
        "clubs": [
            {
                "club": club,
                "players": int(r["players"]),
                "avg_overall": round(float(r["avg_overall"]), 1),
            }
            for club, r in grouped.iterrows()
        ],
    }


# --------------------------------------------------------------------------- #
# 4. Competition queries
# --------------------------------------------------------------------------- #
def standings(
    kb: KnowledgeBase, competition: str, season: int, top_n: int | None = None
) -> dict:
    """Compute a league table from match results for a competition+season.

    Points = 3*W + D. Tie-break: points, goal difference, goals for. Designed
    for round-robin leagues (Série A/B/C); cup/knockout competitions produce a
    table too but it is less meaningful, so callers should prefer brackets.
    """
    comp = resolve_competition(competition)
    df = kb.matches
    df = df[(df["competition"] == comp) & (df["season"] == int(season))]
    df = df[df["home_goal"].notna() & df["away_goal"].notna()]

    # Group by canonical key (a club has several display spellings across
    # files); pick the modal display name for presentation.
    table: dict[str, dict] = {}
    display_counts: dict[str, dict[str, int]] = {}

    def slot(canon: str, display: str) -> dict:
        names = display_counts.setdefault(canon, {})
        names[display] = names.get(display, 0) + 1
        return table.setdefault(
            canon,
            {
                "matches": 0,
                "wins": 0,
                "draws": 0,
                "losses": 0,
                "goals_for": 0,
                "goals_against": 0,
            },
        )

    for _, row in df.iterrows():
        hg, ag = int(row["home_goal"]), int(row["away_goal"])
        h = slot(row["home_canon"], row["home"])
        a = slot(row["away_canon"], row["away"])
        h["matches"] += 1
        a["matches"] += 1
        h["goals_for"] += hg
        h["goals_against"] += ag
        a["goals_for"] += ag
        a["goals_against"] += hg
        if hg > ag:
            h["wins"] += 1
            a["losses"] += 1
        elif ag > hg:
            a["wins"] += 1
            h["losses"] += 1
        else:
            h["draws"] += 1
            a["draws"] += 1

    rows = []
    for canon, entry in table.items():
        entry["team"] = max(display_counts[canon].items(), key=lambda kv: kv[1])[0]
        entry["points"] = entry["wins"] * 3 + entry["draws"]
        entry["goal_difference"] = entry["goals_for"] - entry["goals_against"]
        rows.append(entry)
    rows.sort(
        key=lambda e: (e["points"], e["goal_difference"], e["goals_for"]),
        reverse=True,
    )
    for i, entry in enumerate(rows, start=1):
        entry["position"] = i
    if top_n:
        rows = rows[:top_n]
    return {
        "competition": comp,
        "season": int(season),
        "teams": len(table),
        "standings": rows,
        "champion": rows[0]["team"] if rows else None,
    }


# --------------------------------------------------------------------------- #
# 5. Statistical analysis
# --------------------------------------------------------------------------- #
def competition_stats(
    kb: KnowledgeBase, competition: str | None = None, season: int | None = None
) -> dict:
    """Aggregate goals-per-match, home/away/draw win rates over a filtered set."""
    df = _apply_filters(kb, competition, season)
    df = df[df["home_goal"].notna() & df["away_goal"].notna()]
    n = len(df)
    if n == 0:
        return {
            "competition": competition,
            "season": season,
            "matches": 0,
        }
    hg = df["home_goal"].astype("int64")
    ag = df["away_goal"].astype("int64")
    total_goals = int((hg + ag).sum())
    home_wins = int((hg > ag).sum())
    away_wins = int((ag > hg).sum())
    draws = int((hg == ag).sum())
    return {
        "competition": competition,
        "season": season,
        "matches": n,
        "total_goals": total_goals,
        "avg_goals_per_match": round(total_goals / n, 2),
        "home_wins": home_wins,
        "away_wins": away_wins,
        "draws": draws,
        "home_win_rate": round(100 * home_wins / n, 1),
        "away_win_rate": round(100 * away_wins / n, 1),
        "draw_rate": round(100 * draws / n, 1),
    }


def biggest_wins(
    kb: KnowledgeBase,
    competition: str | None = None,
    season: int | None = None,
    limit: int = 10,
) -> dict:
    """Largest goal-margin victories in the filtered set, margin descending."""
    df = _apply_filters(kb, competition, season)
    df = df[df["home_goal"].notna() & df["away_goal"].notna()].copy()
    df["margin"] = (df["home_goal"].astype("int64") - df["away_goal"].astype("int64")).abs()
    df = df.sort_values(["margin", "match_date"], ascending=[False, False]).head(limit)
    out = []
    for _, row in df.iterrows():
        rec = _match_record(row)
        rec["margin"] = int(row["margin"])
        out.append(rec)
    return {"count": len(out), "matches": out}


def best_record(
    kb: KnowledgeBase,
    competition: str | None = None,
    season: int | None = None,
    venue: str = "either",
    metric: str = "win_rate",
    min_matches: int = 5,
    limit: int = 10,
) -> dict:
    """Rank teams by a record metric (win_rate | points | wins | goal_difference).

    `venue` restricts to home/away matches for "best home/away record" asks.
    Teams with fewer than `min_matches` are excluded to avoid 100%-on-1-game
    noise.
    """
    if metric not in {"win_rate", "points", "wins", "goal_difference"}:
        raise ValueError(f"Unknown metric '{metric}'.")
    df = _apply_filters(kb, competition, season)
    df = df[df["home_goal"].notna() & df["away_goal"].notna()]

    agg: dict[str, dict] = {}
    display_counts: dict[str, dict[str, int]] = {}

    def slot(canon: str, display: str) -> dict:
        names = display_counts.setdefault(canon, {})
        names[display] = names.get(display, 0) + 1
        return agg.setdefault(
            canon,
            {"matches": 0, "wins": 0, "draws": 0, "losses": 0,
             "goals_for": 0, "goals_against": 0},
        )

    for _, row in df.iterrows():
        hg, ag = int(row["home_goal"]), int(row["away_goal"])
        if venue in ("either", "home"):
            e = slot(row["home_canon"], row["home"])
            e["matches"] += 1
            e["goals_for"] += hg
            e["goals_against"] += ag
            if hg > ag:
                e["wins"] += 1
            elif ag > hg:
                e["losses"] += 1
            else:
                e["draws"] += 1
        if venue in ("either", "away"):
            e = slot(row["away_canon"], row["away"])
            e["matches"] += 1
            e["goals_for"] += ag
            e["goals_against"] += hg
            if ag > hg:
                e["wins"] += 1
            elif hg > ag:
                e["losses"] += 1
            else:
                e["draws"] += 1

    rows = []
    for canon, e in agg.items():
        if e["matches"] < min_matches:
            continue
        e["team"] = max(display_counts[canon].items(), key=lambda kv: kv[1])[0]
        e["points"] = e["wins"] * 3 + e["draws"]
        e["goal_difference"] = e["goals_for"] - e["goals_against"]
        e["win_rate"] = round(100 * e["wins"] / e["matches"], 1)
        rows.append(e)
    rows.sort(key=lambda e: (e[metric], e["matches"]), reverse=True)
    return {
        "competition": competition,
        "season": season,
        "venue": venue,
        "metric": metric,
        "teams": rows[:limit],
    }

File: experiments-local/experiment-opencode-brazil/runs/agent=opencode_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep3/src/queries.ts
/**
 * Brazilian Soccer MCP Server - Query Layer
 * -----------------------------------------
 * Context: Pure functions that operate on the in-memory `SoccerData` model
 * produced by `loader.ts`. These functions implement the five required query
 * capabilities from the spec:
 *   1. Match queries (by team, opponent, competition, season, date range)
 *   2. Team queries (per-season stats, home/away splits)
 *   3. Player queries (filter + sort by name, nationality, club, position, rating)
 *   4. Competition queries (computed standings for a season)
 *   5. Statistical analysis (avg goals, biggest wins, best records, head-to-head)
 *
 * Every function is deterministic and side-effect free, which keeps the MCP
 * tool layer thin and the unit tests simple.
 */

import type {
  HeadToHead,
  Match,
  MatchQuery,
  Player,
  PlayerQuery,
  Standing,
  TeamStats,
} from "./types.js";
import type { SoccerData } from "./loader.js";
import { normalizeTeamName } from "./normalize.js";

/** Filter matches by a structured query. Returns most-recent-first. */
export function findMatches(data: SoccerData, q: MatchQuery): Match[] {
  const team = q.team ? normalizeTeamName(q.team) : undefined;
  const opponent = q.opponent ? normalizeTeamName(q.opponent) : undefined;
  const competition = q.competition && q.competition !== "any" ? q.competition : undefined;
  const limit = q.limit && q.limit > 0 ? q.limit : undefined;

  let out = data.matches.filter((m) => {
    if (team) {
      const isHome = m.homeTeam === team;
      const isAway = m.awayTeam === team;
      if (!isHome && !isAway) return false;
      if (opponent) {
        const opp = opponent;
        const oppHome = m.homeTeam === opp;
        const oppAway = m.awayTeam === opp;
        // Need team and opponent both present.
        if (!((isHome && oppAway) || (isAway && oppHome))) return false;
      }
    } else if (opponent) {
      // Only opponent specified -> include matches where opponent plays.
      if (m.homeTeam !== opponent && m.awayTeam !== opponent) return false;
    }
    if (competition && m.competition !== competition) return false;
    if (q.season && m.season !== q.season) return false;
    if (q.startDate && m.date < q.startDate) return false;
    if (q.endDate && m.date > q.endDate) return false;
    return true;
  });

  out = sortMatchesByDateDesc(out);
  if (limit) out = out.slice(0, limit);
  return out;
}

function sortMatchesByDateDesc(matches: Match[]): Match[] {
  return [...matches].sort((a, b) => (a.date < b.date ? 1 : a.date > b.date ? -1 : 0));
}

/** Return all matches between two teams, most-recent-first. */
export function findHeadToHeadMatches(
  data: SoccerData,
  teamA: string,
  teamB: string,
): Match[] {
  const a = normalizeTeamName(teamA);
  const b = normalizeTeamName(teamB);
  const matches = data.matches.filter(
    (m) =>
      (m.homeTeam === a && m.awayTeam === b) ||
      (m.homeTeam === b && m.awayTeam === a),
  );
  return sortMatchesByDateDesc(matches);
}

/** Compute the head-to-head summary between two teams. */
export function headToHead(
  data: SoccerData,
  teamA: string,
  teamB: string,
): HeadToHead {
  const a = normalizeTeamName(teamA);
  const b = normalizeTeamName(teamB);
  const matches = findHeadToHeadMatches(data, teamA, teamB);
  let aWins = 0,
    bWins = 0,
    draws = 0,
    aGoals = 0,
    bGoals = 0;
  for (const m of matches) {
    const aIsHome = m.homeTeam === a;
    const aGoalsIn = aIsHome ? m.homeGoal : m.awayGoal;
    const bGoalsIn = aIsHome ? m.awayGoal : m.homeGoal;
    aGoals += aGoalsIn;
    bGoals += bGoalsIn;
    if (aGoalsIn > bGoalsIn) aWins++;
    else if (bGoalsIn > aGoalsIn) bWins++;
    else draws++;
  }
  return {
    teamA: a,
    teamB: b,
    matches: matches.length,
    teamAWins: aWins,
    teamBWins: bWins,
    draws,
    teamAGoals: aGoals,
    teamBGoals: bGoals,
    recent: matches,
  };
}

/** Compute aggregate stats for a team, optionally restricted by season/competition. */
export function teamStats(
  data: SoccerData,
  team: string,
  opts?: { season?: number; competition?: string },
): TeamStats {
  const t = normalizeTeamName(team);
  const matches = data.matches.filter(
    (m) =>
      (m.homeTeam === t || m.awayTeam === t) &&
      (!opts?.season || m.season === opts.season) &&
      (!opts?.competition || m.competition === opts.competition),
  );
  return computeTeamStats(t, matches);
}

/** Compute stats from an explicit list of matches involving one team. */
function computeTeamStats(team: string, matches: Match[]): TeamStats {
  const home: TeamStats["home"] = {
    matches: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goalsFor: 0,
    goalsAgainst: 0,
  };
  const away: TeamStats["away"] = {
    matches: 0,
    wins: 0,
    draws: 0,
    losses: 0,
    goalsFor: 0,
    goalsAgainst: 0,
  };
  let wins = 0,
    draws = 0,
    losses = 0,
    goalsFor = 0,
    goalsAgainst = 0;

  for (const m of matches) {
    const isHome = m.homeTeam === team;
    const gf = isHome ? m.homeGoal : m.awayGoal;
    const ga = isHome ? m.awayGoal : m.homeGoal;
    goalsFor += gf;
    goalsAgainst += ga;
    let result: "win" | "draw" | "loss";
    if (gf > ga) result = "win";
    else if (ga > gf) result = "loss";
    else result = "draw";
    if (result === "win") wins++;
    else if (result === "draw") draws++;
    else losses++;

    const v = isHome ? home : away;
    v.matches++;
    v.goalsFor += gf;
    v.goalsAgainst += ga;
    if (result === "win") v.wins++;
    else if (result === "draw") v.draws++;
    else v.losses++;
  }
  const points = wins * 3 + draws;
  return {
    team,
    matches: matches.length,
    wins,
    draws,
    losses,
    goalsFor,
    goalsAgainst,
    goalDifference: goalsFor - goalsAgainst,
    points,
    home,
    away,
  };
}

/** Search the FIFA player database. */
export function findPlayers(data: SoccerData, q: PlayerQuery): Player[] {
  let out = data.players.filter((p) => {
    if (q.name) {
      const needle = q.name.toLowerCase();
      if (!p.name.toLowerCase().includes(needle)) return false;
    }
    if (q.nationality) {
      const needle = q.nationality.toLowerCase();
      if (!p.nationality.toLowerCase().includes(needle)) return false;
    }
    if (q.club) {
      const needle = normalizeTeamName(q.club);
      const club = normalizeTeamName(p.club);
      if (!club.includes(needle)) return false;
    }
    if (q.position) {
      const needle = q.position.toLowerCase();
      if (!p.position.toLowerCase().includes(needle)) return false;
    }
    if (q.minOverall !== undefined && p.overall < q.minOverall) return false;
    return true;
  });

  const sortBy = q.sortBy ?? "overall";
  const descending = q.descending ?? true;
  out = out.sort((a, b) => {
    let cmp = 0;
    if (sortBy === "name") cmp = a.name.localeCompare(b.name);
    else cmp = (a[sortBy] as number) - (b[sortBy] as number);
    return descending ? -cmp : cmp;
  });

  const limit = q.limit && q.limit > 0 ? q.limit : 50;
  return out.slice(0, limit);
}

/** Compute standings for a league/competition season. */
export function standings(
  data: SoccerData,
  competition: string,
  season: number,
  limit?: number,
): Standing[] {
  const matches = data.matches.filter(
    (m) => m.competition === competition && m.season === season,
  );
  const teams = new Set<string>();
  for (const m of matches) {
    teams.add(m.homeTeam);
    teams.add(m.awayTeam);
  }
  const rows: Standing[] = [];
  for (const t of teams) {
    const stats = computeTeamStats(
      t,
      matches.filter((m) => m.homeTeam === t || m.awayTeam === t),
    );
    rows.push({
      position: 0,
      team: t,
      played: stats.matches,
      wins: stats.wins,
      draws: stats.draws,
      losses: stats.losses,
      goalsFor: stats.goalsFor,
      goalsAgainst: stats.goalsAgainst,
      goalDifference: stats.goalDifference,
      points: stats.points,
    });
  }
  rows.sort((a, b) => {
    if (b.points !== a.points) return b.points - a.points;
    if (b.wins !== a.wins) return b.wins - a.wins;
    if (b.goalDifference !== a.goalDifference)
      return b.goalDifference - a.goalDifference;
    if (b.goalsFor !== a.goalsFor) return b.goalsFor - a.goalsFor;
    return a.team.localeCompare(b.team);
  });
  rows.forEach((r, i) => (r.position = i + 1));
  return limit ? rows.slice(0, limit) : rows;
}

/** Average goals per match across a (filtered) set of matches. */
export function averageGoals(matches: Match[]): {
  perMatch: number;
  totalMatches: number;
  totalGoals: number;
  homeWinRate: number;
  drawRate: number;
  awayWinRate: number;
} {
  const totalMatches = matches.length;
  if (totalMatches === 0) {
    return {
      perMatch: 0,
      totalMatches: 0,
      totalGoals: 0,
      homeWinRate: 0,
      drawRate: 0,
      awayWinRate: 0,
    };
  }
  let totalGoals = 0,
    homeWins = 0,
    draws = 0,
    awayWins = 0;
  for (const m of matches) {
    totalGoals += m.homeGoal + m.awayGoal;
    if (m.winner === "home") homeWins++;
    else if (m.winner === "away") awayWins++;
    else draws++;
  }
  return {
    perMatch: totalGoals / totalMatches,
    totalMatches,
    totalGoals,
    homeWinRate: homeWins / totalMatches,
    drawRate: draws / totalMatches,
    awayWinRate: awayWins / totalMatches,
  };
}

/** Return the biggest victories (by goal difference) in a set of matches. */
export function biggestWins(matches: Match[], limit = 10): Array<{
  match: Match;
  goalDifference: number;
}> {
  return matches
    .map((m) => ({ match: m, goalDifference: Math.abs(m.homeGoal - m.awayGoal) }))
    .filter((x) => x.goalDifference > 0)
    .sort((a, b) => b.goalDifference - a.goalDifference)
    .slice(0, limit);
}

/** Return the team with the best record over the given matches. */
export function bestTeamRecord(
  data: SoccerData,
  matches: Match[],
  venue?: "home" | "away",
): { team: string; stats: TeamStats } | null {
  const teams = new Set<string>();
  for (const m of matches) {
    if (!venue || venue === "home") teams.add(m.homeTeam);
    if (!venue || venue === "away") teams.add(m.awayTeam);
  }
  let best: { team: string; stats: TeamStats } | null = null;
  for (const t of teams) {
    const teamMatches = matches.filter(
      (m) =>
        (venue === "home" && m.homeTeam === t) ||
        (venue === "away" && m.awayTeam === t) ||
        (!venue && (m.homeTeam === t || m.awayTeam === t)),
    );
    if (teamMatches.length === 0) continue;
    const stats = computeTeamStats(t, teamMatches);
    if (!best || stats.points > best.stats.points) {
      best = { team: t, stats };
    }
  }
  return best;
}

/** Distinct competitions a team has played in. */
export function competitionsForTeam(data: SoccerData, team: string): string[] {
  const t = normalizeTeamName(team);
  const set = new Set<string>();
  for (const m of data.matches) {
    if (m.homeTeam === t || m.awayTeam === t) {
      set.add(m.competition);
    }
  }
  return Array.from(set);
}
